tests/test_asset_fingerprinter.py

```diff
@@ -0,0 +1,86 @@
+        )
+
+    def test_building_file_path(self, mocker):
+        get_file_content_mock = mocker.patch.object(AssetFingerprinter, "get_asset_file_contents")
+        get_file_content_mock.return_value = b"""
+            document.write('Hello world!');
+        """
+        fingerprinter = AssetFingerprinter()
+        fingerprinter.get_url("javascripts/application.js")
+        fingerprinter.get_asset_file_contents.assert_called_with("app/static/javascripts/application.js")
+
+    def test_hashes_are_consistent(self, mocker):
+        get_file_content_mock = mocker.patch.object(AssetFingerprinter, "get_asset_file_contents")
+        get_file_content_mock.return_value = b"""
+            body {
+                font-family: nta;
+            }
+        """
+        asset_fingerprinter = AssetFingerprinter()
+        assert asset_fingerprinter.get_asset_fingerprint(
+            "application.css"
+        ) == asset_fingerprinter.get_asset_fingerprint("same_contents.css")
+
+    def test_hashes_are_different_for_different_files(self, mocker):
+        get_file_content_mock = mocker.patch.object(AssetFingerprinter, "get_asset_file_contents")
+        asset_fingerprinter = AssetFingerprinter()
+        get_file_content_mock.return_value = b"""
+            body {
+                font-family: nta;
+            }
+        """
+        css_hash = asset_fingerprinter.get_asset_fingerprint("application.css")
+        get_file_content_mock.return_value = b"""
+            document.write('Hello world!');
+        """
+        js_hash = asset_fingerprinter.get_asset_fingerprint("application.js")
+        assert js_hash != css_hash
+
+    def test_hash_gets_cached(self, mocker):
+        get_file_content_mock = mocker.patch.object(AssetFingerprinter, "get_asset_file_contents")
+        get_file_content_mock.return_value = b"""
+            body {
+                font-family: nta;
+            }
+        """
+        fingerprinter = AssetFingerprinter()
+        assert fingerprinter.get_url("application.css") == "/static/application.css?418e6f4a6cdf1142e45c072ed3e1c90a"
+        fingerprinter._cache["application.css"] = "a1a1a1"
+        assert fingerprinter.get_url("application.css") == "a1a1a1"
+        fingerprinter.get_asset_file_contents.assert_called_once_with("app/static/application.css")
+
+    def test_without_hash_if_requested(self, mocker):
+        fingerprinter = AssetFingerprinter()
+        assert (
+            fingerprinter.get_url(
+                "application.css",
+                with_querystring_hash=False,
+            )
+            == "/static/application.css"
+        )
+        assert fingerprinter._cache == {}
+
+
+class TestAssetFingerprintWithUnicode:
+    def test_can_read_self(self):
+        "Ralph’s apostrophe is a string containing a unicode character"
+        AssetFingerprinter(filesystem_path="tests/").get_url("test_asset_fingerprinter.py")
```

Comment:
I don't think we are asserting anything in this test?

Reply:
This test checks that the fingerprinter can load a file (in this case the test file itself) containing a non-ascii character (Ralph’s apostrophe). This might not be a problem now we don’t have to support Python 2. But it is a cute test

Reply:
Hah, fair, only the string is not matched with any mock or anything, it's just there on it's own, I don't understand how this works tbh 😳

Reply:
Correct, it’s not doing any mocking. It opens and reads the file at `tests/test_asset_fingerprinter.py`, which contains the string. Just so happens that that file is the same one that contains the test itself.

Reply:
To me it looks like this test isn't testing anything? There is no assertion, no comparison... How would I make it fail?

Reply:
If it couldn’t read the file it would raise an exception (causing the test to fail). You could make it fail by modifying `AssetFingerprinter.get_asset_file_contents` to open the file as ASCII instead of bytes

Reply:
Could you make that explicit by wrapping it in a try/catch, as described by https://miguendes.me/how-to-check-if-an-exception-is-raised-or-not-with-pytest?

Reply:
@quis Ahh I get it now, it scans the whole test file - that's why it contains that string, too.

It took me so long to get what's happening there, because the test structure is very unusual.